models/image.py

```diff
@@ -17,37 +14,43 @@
 
 from pipeline.utils import read_fits_image, save_fits_image_file
 
-from models.base import SeeChangeBase, Base, AutoIDMixin, FileOnDiskMixin, SpatiallyIndexed, FourCorners
+from models.base import (
+    Base,
+    SeeChangeBase,
+    SmartSession,
+    AutoIDMixin,
+    FileOnDiskMixin,
+    SpatiallyIndexed,
+    FourCorners,
+    HasBitFlagBadness,
+    _logger
+)
 from models.exposure import Exposure
 from models.instrument import get_instrument_instance
-from models.psf import PSF
 from models.enums_and_bitflags import (
     ImageFormatConverter,
     ImageTypeConverter,
     image_badness_inverse,
-    data_badness_dict,
-    image_preprocessing_dict,
-    string_to_bitflag,
-    bitflag_to_string,
 )
 
 import util.config as config
 
-image_source_self_association_table = sa.Table(
-    'image_sources',
+# links many-to-many Image to all the Images used to create it
+image_upstreams_association_table = sa.Table(
+    'image_upstreams_association',
     Base.metadata,
-    sa.Column('source_id',
+    sa.Column('upstream_id',
               sa.Integer,
-              sa.ForeignKey('images.id', ondelete="CASCADE", name='image_sources_source_id_fkey'),
+              sa.ForeignKey('images.id', ondelete="CASCADE", name='image_upstreams_association_upstream_id_fkey'),
               primary_key=True),
-    sa.Column('combined_id',
+    sa.Column('downstream_id',
               sa.Integer,
-              sa.ForeignKey('images.id',ondelete="CASCADE", name='image_sources_combined_id_fkey'),
+              sa.ForeignKey('images.id', ondelete="CASCADE", name='image_upstreams_association_downstream_id_fkey'),
               primary_key=True),
 )
```

Comment:
So, rather than having this assocate back to images, it should associate back to something else.

That something else is "zeropoint".  Because the coadd process uses source lists, and WCSes and zeropoints from at least some of the images, we have to know all of (image, source_list, wcs, zeropoint) to figure out what actually went into making this coadded (or subtracted) image.

Because it sounds weird to link to a zeropoint, rename Zeropoint to CalibratedImage.  It will have database fields wordcoordinates_id, provenance_id, and the actual zeropoint stuff (zp, apercor, etc).  Give it convenience method "source_list" that gets the source list from the worldcoordinates the CalibratedImage points to, and "image" that gets the image from that source list.

Then, in this table, upstream_id is actually a foreign key into CalibratedImage.  That way, we've got everything we need to figure out what went into the image combination (be it coadd or subtraction).

Reply:
Why don't we rename it ImageProducts?